drivers/charger/bq2407x.c

```diff
@@ -0,0 +1,215 @@
+
+static int ti_bq2407x_get_prop_status(const struct device *dev, union charger_propval *val)
+{
+	const struct ti_bq2407x_config *config = dev->config;
+	int pg;
+	int ce;
+	int stat1;
+	int stat2;
+
+	pg = gpio_pin_get_dt(&config->pg_pin);
+	if (pg < 0) {
+		return pg;
+	}
+
+	if (!pg) {
+		/* External supply is not present */
+		val->status = CHARGER_STATUS_DISCHARGING;
+		return 0;
+	}
+
+	ce = gpio_pin_get_dt(&config->pg_pin);
+	if (ce < 0) {
+		return ce;
+	}
+
+	if (!ce) {
+		/* Charging is not enabled */
+		val->status = CHARGER_STATUS_NOT_CHARGING;
+		return 0;
+	}
+
+	/*
+	 * The status table is as follows [stat1,stat2]
+	 *   [1,1] CHARGER_STATUS_CHARGING (precharge)
+	 *   [1,0] CHARGER_STATUS_CHARGING (fast charge)
+	 *   [0,1] CHARGER_STATUS_FULL
+	 *   [0,0] CHARGER_STATUS_NOT_CHARGING
+	 */
+	stat1 = gpio_pin_get_dt(&config->stat1_pin);
+	if (stat1 < 0) {
+		return stat1;
+	}
+
+	stat2 = gpio_pin_get_dt(&config->stat2_pin);
+	if (stat2 < 0) {
+		return stat2;
+	}
+
+	if (stat1) {
+		val->status = CHARGER_STATUS_CHARGING;
+	} else if (stat2) {
+		val->status = CHARGER_STATUS_FULL;
+	} else {
+		val->status = CHARGER_STATUS_NOT_CHARGING;
+	}
+
+	return 0;
+}
+
+static int ti_bq2407x_get_property(const struct device *dev, const charger_prop_t prop,
+				   union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ret = ti_bq2407x_get_prop_online(dev, val);
+		break;
+
+	case CHARGER_PROP_STATUS:
+		ret = ti_bq2407x_get_prop_status(dev, val);
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int ti_bq2407x_set_prop_status(const struct device *dev,
+				      const union charger_propval *val)
+
+{
+	const struct ti_bq2407x_config *config = dev->config;
+	int ret;
+
+	switch (val->charge_type) {
+	case CHARGER_STATUS_NOT_CHARGING:
+		/* Disable charging */
+		ret = gpio_pin_set_dt(&config->ce_pin, 0);
+		break;
+
+	case CHARGER_STATUS_CHARGING:
+		/* Enable charging */
+		ret = gpio_pin_set_dt(&config->ce_pin, 1);
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return ret;
+}
+
+static int ti_bq2407x_set_property(const struct device *dev, const charger_prop_t prop,
+				    const union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_STATUS:
+		ret = ti_bq2407x_set_prop_status(dev, val);
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static const struct charger_driver_api ti_bq2407x_api = {
+	.get_property = ti_bq2407x_get_property,
+	.set_property = ti_bq2407x_set_property,
+};
+
+static int ti_bq2407x_init(const struct device *dev)
+{
+	const struct ti_bq2407x_config *config = dev->config;
+	int ret;
+
+	if ((gpio_pin_configure_dt(&config->stat1_pin, 0) < 0) ||
```

Comment:
You should use `gpio_is_ready_dt` before that.

Reply:
added check for all gpios before pin configure